RecoCTPPS/TotemRPLocal/plugins/TotemTimingLocalTrackFitter.cc

```diff
@@ -0,0 +1,155 @@
+
+    TotemTimingDetId tmpId( 0, 1, 0, 0, 0 );
+    tmpId.setArm( detId.arm() );
+    tmpId.setRP( detId.rp() );
+    tmpId.setPlane( detId.plane() );
+    if ( planeActivityMap[tmpId] > maxPlaneActiveChannels_ )
+      continue;
+
+    for ( const auto& hit : vec ) {
+      tmpId.setPlane( 0 );
+      if ( trk_algo_map_.find( tmpId ) != trk_algo_map_.end() )
+        trk_algo_map_.find( tmpId )->second.addHit( hit );
+      else
+        edm::LogWarning("TotemTimingLocalTrackFitter") << "Invalid detId for rechit: arm=" << detId.arm() << ", rp=" << detId.rp();
+    }
+  }
+
+  // retrieves tracks for all hit sets
+  for ( auto& trk_algo_entry : trk_algo_map_ )
+    trk_algo_entry.second.produceTracks( pOut->operator[]( trk_algo_entry.first ) );
+
+  iEvent.put( std::move( pOut ) );
+
+  // remove all hits from the track producers to prepare for the next event
+  for ( auto& trk_algo_entry : trk_algo_map_ )
+    trk_algo_entry.second.clear();
+}
+
+void
+TotemTimingLocalTrackFitter::fillDescriptions( edm::ConfigurationDescriptions& descr )
+{
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>( "recHitsTag", edm::InputTag( "totemTimingRecHits" ) )
+    ->setComment( "input rechits collection to retrieve" );
+  desc.add<int>( "verbosity", 0 )
+    ->setComment( "general verbosity of this module" );
+
+  desc.add<double>( "threshold", 1.5 )
+    ->setComment( "minimal number of rechits to be observed before launching the track recognition algorithm" );
+  desc.add<double>( "thresholdFromMaximum", 0.5 )
+    ->setComment( "threshold relative to hit profile function local maximum for determining the width of the track" );
+  desc.add<double>( "resolution", 0.01 /* mm */ )
+    ->setComment( "spatial resolution on the horizontal coordinate (in mm)" );
+  desc.add<double>( "sigma", 0. )
+    ->setComment( "pixel efficiency function parameter determining the smoothness of the step" );
+  desc.add<double>( "tolerance", 0. /* mm */)
+    ->setComment( "tolerance used for checking if the track contains certain hit" );
+  desc.add<int>( "maxPlaneActiveChannels", 3 /* mm */)
+    ->setComment( "threshold for discriminating noisy planes" );
+
+  desc.add<std::string>( "pixelEfficiencyFunction", "(x>[0]-0.5*[1]-0.05)*(x<[0]+0.5*[1]-0.05)+0*[2]" )
+    ->setComment( "efficiency function for single pixel\n"
+                  "can be defined as:\n"
+                  " * Precise: (TMath::Erf((x-[0]+0.5*([1]-0.05))/([2]/4)+2)+1)*TMath::Erfc((x-[0]-0.5*([1]-0.05))/([2]/4)-2)/4\n"
+                  " * Fast: (x>[0]-0.5*([1]-0.05))*(x<[0]+0.5*([1]-0.05))+((x-[0]+0.5*([1]-0.05)+[2])/[2])*(x>[0]-0.5*([1]-0.05)-[2])*(x<[0]-0.5*([1]-0.05))+(2-(x-[0]-0.5*([1]-0.05)+[2])/[2])*(x>[0]+0.5*([1]-0.05))*(x<[0]+0.5*([1]-0.05)+[2])\n"
+                  " * Legacy: (1/(1+exp(-(x-[0]+0.5*([1]-0.05))/[2])))*(1/(1+exp((x-[0]-0.5*([1]-0.05))/[2])))\n"
+                  " * Default (sigma ignored): (x>[0]-0.5*[1]-0.05)*(x<[0]+0.5*[1]-0.05)+0*[2]\n"
```

Comment:
This seems like a good case for the formula to be derived from the key name
to avoid mistakes in copy-pastes.
E.g. "Precise", "Fast" etc will be converted in construction to an appropriate formula and for an unknown value the code will either throw or try to interpret this as a formula.

You may also want to check https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Switch

Reply:
Very good! Did not know that feature yet :slightly_smiling_face: 
If possible, we would however prefer to keep the full modularity to test against drastically different pixel efficiency functions for our commissioning purposes... But indeed once we converge it would definitely help a lot!